File: T-Bank/Algorithms-and-Data-Structures-2025/10sprint-SegmentTree2/contest/D-Окна/solution.py
import sys

class Event:
    __slots__ = ('x', 'y1', 'y2', 'type_')
    def __init__(self, x: int, y1: int, y2: int, type_: int) -> None:
        self.x: int = x
        self.y1: int = y1
        self.y2: int = y2
        self.type_: int = type_
    
    def __lt__(self, other: 'Event') -> bool:
        if self.x != other.x: return self.x < other.x
        return self.type_ > other.type_

class SegmentTree:
    __slots__ = ('size', 'max_cnt', 'cnt', 'lazy', 'max_pos')
    def __init__(self, n: int) -> None:
        self.size: int = 1
        while self.size < n: self.size <<= 1
        self.max_cnt: list[int] = [0] * (2 * self.size)
        self.cnt: list[int] = [0] * (2 * self.size)
        self.lazy: list[int] = [0] * (2 * self.size)
        self.max_pos: list[int] = [0] * (2 * self.size)
        for i in range(self.size): self.max_pos[i + self.size] = i
        for i in range(self.size - 1, 0, -1): self.max_pos[i] = self.max_pos[2 * i]
    
    def push(self, node: int, node_l: int, node_r: int) -> None:
        if self.lazy[node] == 0: return
        self.max_cnt[node] += self.lazy[node]
        self.cnt[node] += self.lazy[node]
        if node_l != node_r:
            self.lazy[2 * node] += self.lazy[node]
            self.lazy[2 * node + 1] += self.lazy[node]
        self.lazy[node] = 0
    
    def update_range(self, l: int, r: int, v: int, node: int = 1, node_l: int = 0, node_r: int = None) -> None:
        if node_r is None: node_r = self.size - 1
        self.push(node, node_l, node_r)
        if r < node_l or l > node_r: return
        if l <= node_l and node_r <= r:
            self.lazy[node] += v
            self.push(node, node_l, node_r)
            return
        mid: int = node_l + ((node_r - node_l) >> 1)
        self.update_range(l, r, v, 2 * node, node_l, mid)
        self.update_range(l, r, v, 2 * node + 1, mid + 1, node_r)
        if self.max_cnt[2 * node] >= self.max_cnt[2 * node + 1]:
            self.max_cnt[node] = self.max_cnt[2 * node]
            self.max_pos[node] = self.max_pos[2 * node]
        else:
            self.max_cnt[node] = self.max_cnt[2 * node + 1]
            self.max_pos[node] = self.max_pos[2 * node + 1]
        self.cnt[node] = max(self.cnt[2 * node], self.cnt[2 * node + 1])
    
    def get_max(self) -> tuple[int, int]:
        self.push(1, 0, self.size - 1)
        return (self.max_cnt[1], self.max_pos[1])

def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline())
    events: list[Event] = []
    min_y: int = float('inf')
    max_y: int = float('-inf')
    for _ in range(n):
        x1, y1, x2, y2 = map(int, sys.stdin.readline().split())
        if y1 > y2: y1, y2 = y2, y1
        events.append(Event(x1, y1, y2, 1))
        events.append(Event(x2, y1, y2, -1))
        min_y = min(min_y, y1)
        max_y = max(max_y, y2)
    events.sort()
    y_range = max_y - min_y + 1
    st: SegmentTree = SegmentTree(y_range)
    max_overlap: int = 0
    result_x: int = 0
    result_y: int = 0
    for event in events:
        current_max, pos = st.get_max()
        if current_max > max_overlap:
            max_overlap = current_max
            result_x = event.x
            result_y = pos + min_y
        st.update_range(event.y1 - min_y, event.y2 - min_y, event.type_)
    sys.stdout.write('{}\n{} {}\n'.format(max_overlap, result_x, result_y))

if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/10sprint-SegmentTree2/contest/D-Окна/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

struct Event {
    int x, y1, y2, type;
    Event(int x, int y1, int y2, int type) : x(x), y1(y1), y2(y2), type(type) {}
    bool operator<(const Event& other) const {
        if (x != other.x) return x < other.x;
        return type > other.type;
    }
};

struct SegmentTree {
    vector<int> max_cnt, cnt, lazy;
    vector<int> max_pos;
    int size;

    SegmentTree(int n) {
        size = 1;
        while (size < n) size <<= 1;
        max_cnt.resize(2 * size, 0);
        cnt.resize(2 * size, 0);
        lazy.resize(2 * size, 0);
        max_pos.resize(2 * size, 0);
        for (int i = 0; i < size; ++i) {
            max_pos[i + size] = i;
        }
        for (int i = size - 1; i > 0; --i) {
            max_pos[i] = max_pos[2 * i];
        }
    }

    void push(int node, int node_l, int node_r) {
        if (lazy[node] == 0) return;
        max_cnt[node] += lazy[node];
        cnt[node] += lazy[node];
        if (node_l != node_r) {
            lazy[2 * node] += lazy[node];
            lazy[2 * node + 1] += lazy[node];
        }
        lazy[node] = 0;
    }

    void update_range(int l, int r, int val, int node = 1, int node_l = 0, int node_r = -1) {
        if (node_r == -1) node_r = size - 1;
        push(node, node_l, node_r);
        if (r < node_l || l > node_r) return;
        if (l <= node_l && node_r <= r) {
            lazy[node] += val;
            push(node, node_l, node_r);
            return;
        }
        int mid = (node_l + node_r) / 2;
        update_range(l, r, val, 2 * node, node_l, mid);
        update_range(l, r, val, 2 * node + 1, mid + 1, node_r);
        if (max_cnt[2 * node] >= max_cnt[2 * node + 1]) {
            max_cnt[node] = max_cnt[2 * node];
            max_pos[node] = max_pos[2 * node];
        } else {
            max_cnt[node] = max_cnt[2 * node + 1];
            max_pos[node] = max_pos[2 * node + 1];
        }
        cnt[node] = max(cnt[2 * node], cnt[2 * node + 1]);
    }

    pair<int, int> get_max() {
        push(1, 0, size - 1);
        return {max_cnt[1], max_pos[1]};
    }
};

void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    int n;
    cin >> n;
    vector<Event> events;
    int min_y = INT_MAX, max_y = INT_MIN;
    for (int i = 0; i < n; ++i) {
        int x1, y1, x2, y2;
        cin >> x1 >> y1 >> x2 >> y2;
        if (y1 > y2) swap(y1, y2);
        events.emplace_back(x1, y1, y2, 1);
        events.emplace_back(x2, y1, y2, -1);
        min_y = min(min_y, y1);
        max_y = max(max_y, y2);
    }
    sort(events.begin(), events.end());
    int y_range = max_y - min_y + 1;
    SegmentTree st(y_range);
    int max_overlap = 0;
    int result_x = 0, result_y = 0;
    for (const auto& event : events) {
        int current_max = st.get_max().first;
        if (current_max > max_overlap) {
            max_overlap = current_max;
            result_x = event.x;
            result_y = st.get_max().second + min_y;
        }
        st.update_range(event.y1 - min_y, event.y2 - min_y, event.type);
    }
    cout << max_overlap << "\n" << result_x << " " << result_y << "\n";
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    solution();
    return 0;
}